internal/app/machined/pkg/controllers/runtime/extension_services_config_files.go

```diff
@@ -0,0 +1,211 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package runtime
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"go.uber.org/zap"
+
+	v1alpha1runtime "github.com/siderolabs/talos/internal/app/machined/pkg/runtime"
+	"github.com/siderolabs/talos/pkg/machinery/resources/runtime"
+)
+
+// ExtensionServicesConfigFilesController writes down the config files for extension services.
+type ExtensionServicesConfigFilesController struct {
+	V1Alpha1Mode            v1alpha1runtime.Mode
+	V1Alpha1Services        ServiceManager
+	ExtensionsConfigBaseDir string
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *ExtensionServicesConfigFilesController) Name() string {
+	return "runtime.ExtensionServicesConfigFilesController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *ExtensionServicesConfigFilesController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: runtime.NamespaceName,
+			Type:      runtime.ExtensionServicesConfigType,
+			Kind:      controller.InputStrong,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *ExtensionServicesConfigFilesController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *ExtensionServicesConfigFilesController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	if ctrl.V1Alpha1Mode == v1alpha1runtime.ModeContainer {
+		return nil
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		list, err := safe.ReaderListAll[*runtime.ExtensionServicesConfig](ctx, r)
+		if err != nil {
+			return fmt.Errorf("error listing extension services config: %w", err)
+		}
+
+		// add finalizers for all live resources
+		for iter := list.Iterator(); iter.Next(); {
+			if iter.Value().Metadata().Phase() != resource.PhaseRunning {
+				continue
+			}
+
+			if err = r.AddFinalizer(ctx, iter.Value().Metadata(), ctrl.Name()); err != nil {
+				return fmt.Errorf("error adding finalizer: %w", err)
+			}
+		}
+
+		touchedDirs := map[string]struct{}{}
+
+		for iter := list.Iterator(); iter.Next(); {
+			extensionConfigPath := filepath.Join(ctrl.ExtensionsConfigBaseDir, iter.Value().Metadata().ID())
+
+			switch iter.Value().Metadata().Phase() {
+			case resource.PhaseTearingDown:
```

Comment:
this feels really complicated, and hard to test properly

I haven't read the whole change, but I assume we say that Talos owns all files for the extensions configuration, so in that case we could simply remove any file which shouldn't be there as part of any reconciliation (and that allows to skip finalizers)

Reply:
hmm that might be easier, i went down the path of checking the teraingdownphase since even a file removal needs a restart of the extension service

Reply:
makes sense... can we detect which services were affected from the paths? as we update/remove things?

Reply:
kind of, from the parent folder, but it just seemed brittle and if someone manually creates a folder, the code would try to restart a non-existing service,

Reply:
but i guess maybe we can still remove files we didn't touch and clean up part of code

Reply:
as it's `tmpfs`, we still know which parts we touched, as it can't be leftover from previous runs, as the controller runs in can keep the mapping of paths and which extensions they are related to